Topic attribute bound electrons

n01r opened the pull request.

n01r commented:
Hey @ax3l and @psychocoderHPC again,
This PR is the continuation of my previous PR #6 and is based on that branch. 
It changes the attribute `chargeState` to `boundElectrons`.
Feel free to comment!

ax3l commented on src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSI.hpp:
```diff
@@ -56,13 +58,16 @@ namespace ionization
             {
                 /* set new particle charge state */
-                parentIon[chargeState_] = 1 + parentIon[chargeState_];
+                parentIon[boundElectrons_] = parentIon[boundElectrons_] - 1;
```
1 is `int`, use `float_X(1.0)`

ax3l commented on src/picongpu/include/particles/traits/GetProtonNumber.hpp:
```diff
@@ -33,7 +33,7 @@ struct GetProtonNumber
     typedef typename PMacc::traits::Resolve<
         typename GetFlagType<typename T_Species::FrameType,protonNumber<> >::type 
-        >::type type;
+    >::type type;
```
unnecessary change, revert pls.

ax3l commented on src/picongpu/include/simulation_defines/param/speciesAttributes.param:
```diff
@@ -55,13 +55,13 @@ value_identifier(float_X, weighting, 0.0);
- * - \todo maybe calculate from new attribute boundElectrons */
-value_identifier(float_X,chargeState,float_X(1.0));
+ */
+value_identifier(float_X,boundElectrons,float_X(1.0));
```
Probably `1.0` is not the best "general" default value.
Maybe "fully ionized" aka `float_X(0.0)` is a better choice for now (or later on and alternatively, the proton-number for "neutral").

ax3l commented on src/picongpu/include/traits/attribute/GetChargeState.hpp:
```diff
@@ -0,0 +1,88 @@
+    template<typename T_Particle>
+    HDINLINE float_X operator()(const T_Particle& particle)
+    {
+        return 1.0;
```
? the default should be zero, too.

ax3l replied:
(pls stop assuming "the world is only neutral hydrogen" while coding ;) )

n01r replied:
That is not what I'm doing. The thing is that as long as ions get cloned out of electrons, I cannot start a simulation with neutral particles. If that particle has `boundElectrons` then I can calculate a `chargeState`. If it doesn't have `boundElectrons` it is most likely an electron and thus singly charged. I could also put an error message in there. "It is no use to ask for a charge state of particles that have no bound electrons because then they are probably elementary particles and their charge does not change. But if you still want to: default is 1 because `charge = CHARGE_CONST * chargeState`." The whole thing is just too preliminary. 
What should happen if you try to call `getChargeState` on a particle without differing charge states?

ax3l replied:
I understand your confusion, but as explained via mail:

> The thing is that as long as ions get cloned out of electrons ...

that is due to historic reasons, feel free to change that to what I described you: initialize all ions first, then clone the electrons from them.

n01r replied:
As I understood it yesterday @psychocoderHPC already has an unproposed branch that allows exactly this. I don't want to meddle in it before he starts a PR.

ax3l replied:
that's fine so simply wait for that :)

else, it's a trivial change in `src/picongpu/include/simulation_defines/param/species*.param`

ax3l commented:
that's fine so simply wait for that :)

else, it's a trivial change in `src/picongpu/include/simulation_defines/param/species*.param`

psychocoderHPC commented:
A general question: 
I don't know if is plays any roll but is `freeElectrons` a more native attribute instead of `boundElectrons`?
The advantage of `freeElectrons` is that for all neutral materials the default is zero.

n01r commented:
@psychocoderHPC 
If you would rename it like that it is the same as `chargeState` just with a different name. That would be zero, too for all neutral materials. It is a bit more native than `chargeState` but as we agreed on before, `boundElectrons` can be extended easier to coding what energy states the bound electrons are actually in.

ax3l commented:
FreeElectrons does not make sence since these e- are not associated with the ion any more. So no :)

psychocoderHPC commented:
yes this make sense

ax3l committed:
Merge pull request #687 from n01r/topic-flagProtonNeutronNumber

Flag for proton and neutron number

Commit:
Remove attribute chargeState, add boundElectrons

In this commit the following has been refactored:
    -- remove attribute `chargeState`, add `boundElectrons`
    -- add trait `GetChargeState.hpp`
    -- replace all uses of `chargeState` with `boundElectrons`
